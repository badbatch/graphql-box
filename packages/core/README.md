# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
