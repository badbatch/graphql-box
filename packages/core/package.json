{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}