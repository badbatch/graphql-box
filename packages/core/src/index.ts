import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
