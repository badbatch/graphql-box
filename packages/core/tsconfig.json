{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}