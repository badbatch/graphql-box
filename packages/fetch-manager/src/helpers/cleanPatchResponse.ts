import { MaybeRawFetchData, PlainObjectMap } from "@graphql-box/core";
import { isObjectLike, keys } from "lodash";

const convertNullArrayEntriesToUndefined = (data: PlainObjectMap) => {
  const converter = (objectLike: PlainObjectMap | any[]) => {
    const isArray = Array.isArray(objectLike);

    return keys(objectLike).reduce(
      (acc: PlainObjectMap, key) => {
        let value;

        if (isArray) {
          const index = Number(key);
          value = objectLike[index] === null ? undefined : objectLike[index];
        } else {
          value = objectLike[key];
        }

        if (isObjectLike(value)) {
          acc[key] = converter(value);
        } else {
          acc[key] = value;
        }

        return acc;
      },
      isArray ? [] : {},
    );
  };

  return converter(data);
};

export default (responseData: MaybeRawFetchData) => {
  const { data, paths } = responseData;

  if (data && paths) {
    return {
      ...responseData,
      data: convertNullArrayEntriesToUndefined(data),
    };
  }

  return responseData;
};
