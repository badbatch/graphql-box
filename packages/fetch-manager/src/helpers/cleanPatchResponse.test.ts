import { type PartialRawFetchData } from '@graphql-box/core';
import { expect } from '@jest/globals';
import { responseDataSets } from '../__testUtils__/responseDataSets.ts';
import { cleanPatchResponse } from './cleanPatchResponse.ts';

describe('cleanPatchResponse', () => {
  it('should merge the response data correctly', () => {
    expect(cleanPatchResponse(responseDataSets[5] as unknown as PartialRawFetchData)).toMatchSnapshot();
  });
});
