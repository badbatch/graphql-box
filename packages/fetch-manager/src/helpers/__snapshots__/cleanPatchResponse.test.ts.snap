// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanPatchResponse should merge the response data correctly 1`] = `
Object {
  "_cacheMetadata": Object {
    "query": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.movie.releaseDates": Object {
      "cacheControl": Object {
        "maxAge": 28800,
        "public": true,
      },
      "etag": "W/\\"570d27346fc74b9c6bbc404f6c60c5a5\\"",
      "ttl": 1645814743457,
    },
    "query.movie.videos": Object {
      "cacheControl": Object {
        "maxAge": 28800,
        "public": true,
      },
      "etag": "W/\\"ef7dc264044a1de7affcf190e991a577\\"",
      "ttl": 1645814743204,
    },
    "query.search": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges.node": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges.node.contentRatings": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges.node.releaseDates": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges.node.releaseDates.releaseDates": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
    "query.search.edges.node.videos": Object {
      "cacheControl": Object {
        "maxAge": 600,
        "public": true,
      },
      "ttl": 1645786542679,
    },
  },
  "data": Object {
    "search": Object {
      "edges": Array [
        undefined,
        Object {
          "node": Object {
            "releaseDates": Array [
              Object {
                "iso_3166_1": "US",
                "releaseDates": Array [
                  Object {
                    "certification": "PG",
                  },
                ],
              },
              Object {
                "iso_3166_1": "GB",
                "releaseDates": Array [
                  Object {
                    "certification": "PG",
                  },
                ],
              },
              Object {
                "iso_3166_1": "DK",
                "releaseDates": Array [
                  Object {
                    "certification": "15",
                  },
                ],
              },
              Object {
                "iso_3166_1": "NL",
                "releaseDates": Array [
                  Object {
                    "certification": "12",
                  },
                  Object {
                    "certification": "12",
                  },
                ],
              },
              Object {
                "iso_3166_1": "BR",
                "releaseDates": Array [
                  Object {
                    "certification": "14",
                  },
                ],
              },
              Object {
                "iso_3166_1": "PT",
                "releaseDates": Array [
                  Object {
                    "certification": "M/12",
                  },
                ],
              },
              Object {
                "iso_3166_1": "ES",
                "releaseDates": Array [
                  Object {
                    "certification": "16",
                  },
                ],
              },
              Object {
                "iso_3166_1": "DE",
                "releaseDates": Array [
                  Object {
                    "certification": "16",
                  },
                ],
              },
              Object {
                "iso_3166_1": "FR",
                "releaseDates": Array [
                  Object {
                    "certification": "U",
                  },
                ],
              },
            ],
          },
        },
      ],
    },
  },
  "hasNext": true,
  "paths": Array [
    "search.edges.1.node",
  ],
  "requestID": "0e9ad890-9628-11ec-9046-412e5b96757a",
}
`;
